applications/nrf5340_audio/src/bluetooth/bt_rendering_and_capture/volume/bt_vol_ctlr.c

```diff
@@ -70,49 +98,149 @@ static bool vcs_client_peer_exists(struct bt_conn *conn)
+ * @param include_caller[in]	If true, also send an update to the initiating VCS.
  */
-static void vcs_state_ctlr_cb_handler(struct bt_vcp_vol_ctlr *vcs, int err, uint8_t volume,
-				      uint8_t mute)
+static void update_volume_of_all_peers(struct bt_vcp_vol_ctlr const *const vcs, uint8_t volume_out,
+				       bool include_caller)
 {
 	int ret;
 
+	for (size_t i = 0; i < ARRAY_SIZE(vcs_client_peer); i++) {
+		LOG_DBG("Saving vol %u for dev %u", volume_out, i);
+		vcs_client_peer[i].master_vol = volume_out;
+		vcs_client_peer[i].unknown_vol = false;
+
+		if (!vcs_peer_alive(vcs_client_peer[i].peer)) {
+			LOG_DBG("Peer %u is not alive", i);
+			continue;
+		}
+
+		if (vcs == vcs_client_peer[i].peer && !include_caller) {
+			LOG_DBG("Caller not included");
+			continue;
+		}
+
+		ret = bt_vcp_vol_ctlr_set_vol(vcs_client_peer[i].peer, volume_out);
+
+		if (ret) {
+			LOG_ERR("Vol sync failed to remote device %u, err = %d", i, ret);
+		}
+	}
+}
+
+static void vol_set_cb_handler(struct bt_vcp_vol_ctlr *vol_ctlr, int err)
+{
+	int idx;
+
+	if (err) {
+		LOG_WRN("VCS vol set callback error: %d", err);
+		return;
+	}
+
+	idx = vcs_peer_index_get(vol_ctlr);
+	if (idx < 0) {
+		LOG_ERR("VCS peer not found");
+		return;
+	}
+
+	LOG_DBG("Saving vol %u for dev %u", vol_ctlr->cp_val.volume, idx);
+	vcs_client_peer[idx].master_vol = vol_ctlr->cp_val.volume;
+	vcs_client_peer[idx].unknown_vol = false;
+}
+
+static void vcs_state_cb_handler(struct bt_vcp_vol_ctlr *vcs, int err, uint8_t dev_volume,
+				 uint8_t mute)
+{
+	int ret;
+	int current_idx = vcs_peer_index_get(vcs);
+
 	if (err) {
 		LOG_ERR("VCS state callback error: %d", err);
 		return;
 	}
 
-	for (int i = 0; i < ARRAY_SIZE(vcs_client_peer); i++) {
-		if (vcs == vcs_client_peer[i]) {
-			LOG_DBG("VCS state from remote device %d:", i);
-			continue;
-		}
+	if (current_idx < 0) {
+		LOG_ERR("VCS peer not found");
+		return;
+	}
 
-		LOG_DBG("Sync with other devices %d", i);
+	LOG_INF("VCS state ctrl cb idx: %u vol: %u mute: %u", current_idx, dev_volume, mute);
 
-		if (vcs_client_peer[i] == NULL) {
-			/* Skip */
-			continue;
+	if (vcs_client_peer[current_idx].unknown_vol) {
+		LOG_DBG("VCS default state from device: %u", current_idx);
+		uint8_t common_vol;
+
+		ret = common_vol_get();
+		if (ret < 0) {
+			LOG_DBG("Saving vol %u for renderer %u", dev_volume, current_idx);
+			vcs_client_peer[current_idx].master_vol = dev_volume;
+			vcs_client_peer[current_idx].unknown_vol = false;
+			return;
 		}
 
-		ret = bt_vcp_vol_ctlr_set_vol(vcs_client_peer[i], volume);
-		if (ret) {
-			LOG_DBG("Failed to sync volume to remote device %d, err = "
-				"%d",
-				i, ret);
+		common_vol = ret;
+
+		if (common_vol == dev_volume) {
+			/* Volume of this device is the same as the common volume. No update */
+			return;
 		}
+
+		LOG_DBG("Stored vol found. Writing vol %u to remote device %u", common_vol,
+			current_idx);
+		update_volume_of_all_peers(vcs, common_vol, true);
```

Comment:
How well does this handle situations where multiple servers change volume at the same time? e.g. if you are connected to 2 VCS devices and they both change their volume to different values:

1) Server 1 and Server 2 have volume 100
2) Server 1 locally changes volume to 110 and sends notification, which triggers `update_volume_of_all_peers` to update volume to 110 for Server 2
3) Server 2 locally changes volume to 90 and sends notification before getting the volume request sent in 2), which triggers `update_volume_of_all_peers` to update volume to 90 for Server 1
4) Server 2 now gets volume request for 110, and sends notification for 110, but now you are setting volume to 90 for Server 1, and that triggers yet another call to `update_volume_of_all_peers`

In the above case, won't you effectively end up in an infinitely loop of setting volume between 90 and 110 because they will never really sync?

Reply:
Interesting observation. When tested, I get -EBUSY from `bt_vcp_vol_ctlr_set_vol`

Reply:
> Interesting observation. When tested, I get -EBUSY from `bt_vcp_vol_ctlr_set_vol`

Likely because the previous command has not yet finished for that connection. So depending on timing it may work, it may run in the above loop or it may return an error. To me that sounds like 1/3 cases are handled (although the remaining 2 unhandled cases are rare).